### PR TITLE
(WIP) Tie-breaking extensibility

### PR DESCRIPTION
**Pull request by Nzteb**
Tie-breaking options/documentation, see #96

### CODE REVIEW

File: kge/config-default.yaml [hunk 1]
```diff
@@ -359,6 +359,21 @@ eval:
   # Other options
   pin_memory: False
 
+entity_ranking:
```
Comment (rgemulla) [inline kge/config-default.yaml:362]: Move this under eval for now. It's a good idea to have separate keys in general, but also other things would need to be moved (e.g., the filter splits). We can do this later in a separate PR.

File: kge/config-default.yaml [hunk 2]
```diff
@@ -359,6 +359,21 @@ eval:
...
+  # How to handle cases where the correct query answer has the same score as one or
+  # multiple wrong query answers, for example: Query: (s, p, ?).  (Filtered) Answers:
+  # ['a': score 10, 'b': score 10, 'c': score 10, 'd': score 11]. Correct query
```
Comment (rgemulla) [inline kge/config-default.yaml:365]: The example should have one answer above and one below the 10's.

File: kge/config-default.yaml [hunk 3]
```diff
@@ -359,6 +359,21 @@ eval:
...
+  # answer: 'a'. Possible options are:
+  # - rounded_mean_rank: Calculate the rounded mean rank. The final rank is calculated
+  #                      as 1 + #scores_higher + #answers_with_same_score // 2,
```
Comment (rgemulla) [inline kge/config-default.yaml:368]: write as round_up(...), use /2. Easier to read this way.

File: kge/config-default.yaml [hunk 4]
```diff
@@ -359,6 +359,21 @@ eval:
...
+  #                      where "//" denotes the floor-division. In the example, the
+  #                      final rank for correct answer (s, p,'a') is 1 + 1 + 3 // 2 = 3.
+  # - top:               Assign the best possible rank for the correct answer. The
+  #                      final rank for answer (s, p, 'a') is 2.
```
Comment (rgemulla) [inline kge/config-default.yaml:372]: Write: DO NOT USE THIS, it leads to misleading evaluation results.

File: kge/config-default.yaml [hunk 5]
```diff
@@ -359,6 +359,21 @@ eval:
...
+  # - bottom:            Assign the worst possible rank for the correct answer. The
+  #                      final rank for answer (s, p, 'a') is 4.
```
Comment (rgemulla) [inline kge/config-default.yaml:374]: worst possible - highest. Key name is confusing (bottom for highest rank). Perhaps: worst_rank. LIkewise top -> best_rank.

File: kge/job/entity_ranking.py [hunk 6]
```diff
@@ -44,6 +47,17 @@ def _prepare(self):
             pin_memory=self.config.get("eval.pin_memory"),
         )
 
+        # assign a tie breaker
+        tie_handling = self.config.get("entity_ranking.tie_handling")
```
Comment (rgemulla) [inline kge/job/entity_ranking.py:51]: store this in self.tie_handling

File: kge/job/entity_ranking.py [hunk 7]
```diff
@@ -44,6 +47,17 @@ def _prepare(self):
...
+        if tie_handling == "rounded_mean_rank":
```
Comment (rgemulla) [inline kge/job/entity_ranking.py:52]: And move this to get_ranks. Gives us more flexibility later on, if we want to use better strategies (such as mean_metric).

File: kge/config-default.yaml (hunk 3 above)
Comment (Nzteb) [reply kge/config-default.yaml]: I have addressed the review. Only here I used round_down() as this matches X//Y.

File: kge/config-default.yaml [hunk 8]
```diff
@@ -330,6 +330,23 @@ eval:
...
   type: entity_ranking
 
+  entity_ranking:
+      # How to handle cases where the correct query answer has the same score as one or
+      # multiple wrong query answers. For example: Query: (s, p, ?). (Filtered) Answers:
+      # ['a': score 10, 'b': score 10, 'c': score 10, 'd': score 11, 'e': score 9].
+      # Correct query answer: 'a'. Possible options are:
+      # - rounded_mean_rank: Calculate the rounded mean rank. The final rank is
+      #                      calculated as 1 + #scores_higher +
+      #                      round_down(#answers_with_same_score / 2). In the example,
```
Comment (rgemulla) [inline kge/config-default.yaml:340]: We always round up to err on the worst_rank side.

File: kge/config-default.yaml [hunk 9]
```diff
@@ -330,6 +330,23 @@ eval:
...
+      #                      the final rank for correct answer (s, p,'a') is
+      #                      1 + 1 + round_down(3 / 2) = 3.
+      # - best_rank:         Assign the lowest rank for the correct answer. The final
+      #                      rank for answer (s, p, 'a') is 2. We do not recommend to
+      #                      use this option, it leads to misleading results.
```
Comment (rgemulla) [inline kge/config-default.yaml:345]: To weak. Write DO NOT USE THIS OPTION. It's the reason for a number of problematic results/papers

File: kge/job/entity_ranking.py [hunk 10]
```diff
@@ -13,6 +13,11 @@ class EntityRankingJob(EvaluationJob):
 
     def __init__(self, config: Config, dataset: Dataset, parent_job, model):
         super().__init__(config, dataset, parent_job, model)
+        self.config.check(
+            "eval.entity_ranking.tie_handling",
```
Comment (rgemulla) [inline kge/job/entity_ranking.py:17]: Why not: eval.tie_handling. It makes sense for all ranking tasks, even the ones we still plan to implement

Comment (Nzteb) [reply kge/job/entity_ranking.py:17]: Yes that also makes sense. So dropping the key 'entity_ranking' completely for now.

File: kge/config-default.yaml (hunk 8 above)
Comment (rgemulla) [reply kge/config-default.yaml:340]: Ahh, I read this as "other_answers_with_same_score". Prehaps clarifiy: rounded (up) mean rank,

Comment (Nzteb) [reply kge/config-default.yaml:340]: I'm just describing what the code does.  And in ranks + num_ties // 2 ,  "//" rounds the rank down?

Comment (rgemulla) [reply kge/config-default.yaml:340]: See, you are confused as well. We round the number of ties (>=1) down, which corresponds to rounding the mean rank up. E.g., if there is **one** other answer with he same score than the correct one, then num_ties=**2**. We round down num_ties/2, but this corresponds to rounding up the mean rank. Perhaps its best to not show this computation here, it's only conusing. Mean rank rounded up is clear enough.

File: kge/config-default.yaml [hunk 11]
```diff
@@ -330,6 +330,22 @@ eval:
...
+  # Possible options are:
+  # - worst_rank:        Use the highest rank of all answers that have the same
+  #                      score as the correct answer. In example: 4.
+  # - best_rank:         Use the lowest rank of all answers that have the same
+  #                      score as the correct answer (competition scoring). In
+  #                      example: 2. DO NOT USE THIS OPTION, it leads to
+  #                      misleading evaluation results.
+  # - rounded_mean_rank: Average between worst and best rank, rounded up
+  #                      (rounded fractional ranking). In example: 3.
+  tie_handling: rounded_mean_rank
```
Comment (rgemulla) [inline kge/config-default.yaml:347]: How about this?

**Comment by rgemulla** [comment]
Thanks!

File: kge/config-default.yaml (hunk 11 above)
Comment (Nzteb) [reply kge/config-default.yaml:347]: Cool!

**Pull request merged**